infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -283,6 +328,24 @@ private LogReplicationSession constructSession(String sourceClusterId, String si
                 .build();
     }
 
+    /**
+     * When a node acquires the lock, check if all the sessions present in the system tables are still valid.
+     *
+     * Since the metadata table is written to only by the leader node, we can run into the scenario where the leader node
+     * missed updating the system table because the leadership changed at that precise moment.
+     * If the new leader node had already received the topology change notification before becoming the leader, the stale
+     * sessions would not be recognized on "refresh".
+     * Therefore, iterate over the sessions in the system tables and remove the stale sessions if any.
+     */
+    public void removeStaleSessionOnLeadershipAcquire() {
```

Comment:
True, that is what is happening...We have the in-memory sessions data on all the nodes, but only the leader writes to the metadata table.

Anny pointed out a gap in this: say the leader node is the last one to get the topology notification and just before it could write to the metadata table, leadership changes.
Now, the new leader already has the new topology, and refresh will not find any session to be added/removed. We then have stale data remaining in our system tables.
To keep it clean, we now need to go through one our system tables and ensure that all the sessios in that table are valid. Any invalid sessions will get removed from the system tables. This is done everytime a node acquires leadership

Reply:
Adding this gap in comment section seems useful.

Reply:
this "gap" is already explained in the comment.